src/middlewares/rateLimit.ts

```diff
@@ -0,0 +1,58 @@
+import { Request, Response, NextFunction } from 'express'
+import { RequestersList } from '../utils'
+import { CONFIG as config } from '../config'
+import { sleep } from '../utils'
+import blackList from '../../blacklist.json'
+import spammerList from '../../spammerlist.json'
+
+// Export for testing
+export const requestersList = new RequestersList(blackList, spammerList)
+
+interface RpcRequest {
+  method: string
+  params: any[]
+}
+
+async function handleRejection(res: Response, softReject: boolean): Promise<void> {
+  if (softReject) {
+    const randomSleepTime = 10 + Math.floor(Math.random() * 10)
+    await sleep(randomSleepTime * 1000)
+    res.status(503).send('Network is currently busy. Please try again later.')
+  } else {
+    res.status(429).send('Rejected by rate-limiting')
+  }
+}
+
+async function checkRequest(ip: string, request: RpcRequest): Promise<boolean> {
+  return await requestersList.isRequestOkay(ip, request.method, request.params)
+}
+
+export async function rateLimitMiddleware(req: Request, res: Response, next: NextFunction) {
+  if (!config.rateLimit) {
+    next()
+    return
+  }
+  let ip = String(req.socket.remoteAddress)
+  if (ip.substring(0, 7) == '::ffff:') {
+    ip = ip.substring(7)
+  }
+
+  const requests: RpcRequest[] = Array.isArray(req.body) ? req.body : [req.body]
+
+  try {
+    const results = await Promise.all(
```

Comment:
consider using Promise.allSettled instead.. because if any one of the request's promise  in batch request is rejected then the entire operation will fail

Reply:
as mentioned [here](https://github.com/shardeum/json-rpc-server/pull/91/files#diff-e38932e57dff8f36e0057222569bbcbf31a50d9578cd586efb49076e443391eaR47). the batch rejection is intentional